File: kapeko-track-b-complete/backend/src/main/java/com/song/kapeko/service/LlmService.java
package com.song.kapeko.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;
import java.util.Map;

/**
 * LlmService — provider-agnostic LLM client.
 * Supports: Azure OpenAI | OpenAI | Groq | Anthropic
 *
 * All Level 2, 3, and 4 services call through here.
 * Switch provider via application.properties:
 *   llm.provider=azure | openai | groq | anthropic
 */
@Service
public class LlmService {

    private final Environment env;
    private final ObjectMapper mapper = new ObjectMapper();

    public LlmService(Environment env) {
        this.env = env;
    }

    // ── complete() — single-turn, used by Labs 3, 4, 7, 9, 10 ────────────

    public String complete(String systemPrompt, String userPrompt) {
        String provider = env.getProperty("llm.provider", "azure");
        return switch (provider.toLowerCase()) {
            case "anthropic" -> completeAnthropic(systemPrompt, userPrompt);
            case "groq"      -> completeOpenAICompat("https://api.groq.com", systemPrompt, userPrompt);
            case "openai"    -> completeOpenAICompat("https://api.openai.com", systemPrompt, userPrompt);
            default          -> completeAzure(systemPrompt, userPrompt);
        };
    }

    // ── completeWithHistory() — multi-turn, used by Lab 6 Shopping Copilot ─

    public String completeWithHistory(String systemPrompt, List<Map<String, String>> history) {
        String provider = env.getProperty("llm.provider", "azure");
        return switch (provider.toLowerCase()) {
            case "anthropic" -> completeAnthropicWithHistory(systemPrompt, history);
            case "groq"      -> completeOpenAICompatWithHistory("https://api.groq.com", systemPrompt, history);
            case "openai"    -> completeOpenAICompatWithHistory("https://api.openai.com", systemPrompt, history);
            default          -> completeAzureWithHistory(systemPrompt, history);
        };
    }

    // ── embed() — vector embedding, used by Labs 7 and 8 ────────────────────

    public float[] embed(String text) {
        String provider = env.getProperty("llm.provider", "azure");
        return switch (provider.toLowerCase()) {
            case "openai" -> embedOpenAI(text);
            case "groq"   -> embedFallback(text); // Groq has no embedding API
            case "anthropic" -> embedFallback(text); // Use embedFallback or switch to OpenAI for embeddings
            default       -> embedAzure(text);
        };
    }

    // ── Azure OpenAI ─────────────────────────────────────────────────────────

    private String completeAzure(String system, String user) {
        ObjectNode body = mapper.createObjectNode();
        body.put("temperature", 0.7);
        ArrayNode messages = body.putArray("messages");
        messages.addObject().put("role", "system").put("content", system);
        messages.addObject().put("role", "user").put("content", user);
        String response = azureClient().post()
                .uri("/chat/completions?api-version=2024-02-01")
                .bodyValue(body.toString()).retrieve().bodyToMono(String.class).block();
        return extractContent(response);
    }

    private String completeAzureWithHistory(String system, List<Map<String, String>> history) {
        ObjectNode body = mapper.createObjectNode();
        body.put("temperature", 0.7);
        ArrayNode messages = body.putArray("messages");
        messages.addObject().put("role", "system").put("content", system);
        history.forEach(m -> messages.addObject().put("role", m.get("role")).put("content", m.get("content")));
        String response = azureClient().post()
                .uri("/chat/completions?api-version=2024-02-01")
                .bodyValue(body.toString()).retrieve().bodyToMono(String.class).block();
        return extractContent(response);
    }

    private float[] embedAzure(String text) {
        // Embedding uses a different deployment — text-embedding-ada-002
        String embeddingEndpoint = env.getProperty("llm.azure.embedding-endpoint",
                env.getProperty("llm.azure.endpoint", "").replace(
                        env.getProperty("llm.model", "gpt-4o"), "text-embedding-ada-002"));
        ObjectNode body = mapper.createObjectNode();
        body.put("input", text);
        String response = WebClient.builder().baseUrl(embeddingEndpoint)
                .defaultHeader("api-key", env.getProperty("llm.api-key"))
                .defaultHeader("Content-Type", "application/json")
                .build().post()
                .uri("/embeddings?api-version=2024-02-01")
                .bodyValue(body.toString()).retrieve().bodyToMono(String.class).block();
        return parseEmbedding(response);
    }

    private WebClient azureClient() {
        return WebClient.builder()
                .baseUrl(env.getProperty("llm.azure.endpoint", ""))
                .defaultHeader("api-key", env.getProperty("llm.api-key"))
                .defaultHeader("Content-Type", "application/json")
                .build();
    }

    // ── OpenAI-compatible (OpenAI + Groq use same API format) ────────────────

    private String completeOpenAICompat(String baseUrl, String system, String user) {
        ObjectNode body = mapper.createObjectNode();
        body.put("model", env.getProperty("llm.model", "gpt-4o"));
        body.put("temperature", 0.7);
        ArrayNode messages = body.putArray("messages");
        messages.addObject().put("role", "system").put("content", system);
        messages.addObject().put("role", "user").put("content", user);
        String response = openAICompatClient(baseUrl).post()
                .uri(baseUrl.contains("groq") ? "/openai/v1/chat/completions" : "/v1/chat/completions")
                .bodyValue(body.toString()).retrieve().bodyToMono(String.class).block();
        return extractContent(response);
    }

    private String completeOpenAICompatWithHistory(String baseUrl, String system, List<Map<String, String>> history) {
        ObjectNode body = mapper.createObjectNode();
        body.put("model", env.getProperty("llm.model", "gpt-4o"));
        body.put("temperature", 0.7);
        ArrayNode messages = body.putArray("messages");
        messages.addObject().put("role", "system").put("content", system);
        history.forEach(m -> messages.addObject().put("role", m.get("role")).put("content", m.get("content")));
        String response = openAICompatClient(baseUrl).post()
                .uri(baseUrl.contains("groq") ? "/openai/v1/chat/completions" : "/v1/chat/completions")
                .bodyValue(body.toString()).retrieve().bodyToMono(String.class).block();
        return extractContent(response);
    }

    private float[] embedOpenAI(String text) {
        ObjectNode body = mapper.createObjectNode();
        body.put("model", "text-embedding-ada-002");
        body.put("input", text);
        String response = openAICompatClient("https://api.openai.com").post()
                .uri("/v1/embeddings").bodyValue(body.toString())
                .retrieve().bodyToMono(String.class).block();
        return parseEmbedding(response);
    }

    private WebClient openAICompatClient(String baseUrl) {
        return WebClient.builder().baseUrl(baseUrl)
                .defaultHeader("Authorization", "Bearer " + env.getProperty("llm.api-key"))
                .defaultHeader("Content-Type", "application/json")
                .build();
    }

    // ── Anthropic ─────────────────────────────────────────────────────────────

    private String completeAnthropic(String system, String user) {
        ObjectNode body = mapper.createObjectNode();
        body.put("model", env.getProperty("llm.model", "claude-sonnet-4-6"));
        body.put("max_tokens", 1024);
        body.put("system", system);
        body.putArray("messages").addObject().put("role", "user").put("content", user);
        String response = WebClient.builder().baseUrl("https://api.anthropic.com")
                .defaultHeader("x-api-key", env.getProperty("llm.api-key"))
                .defaultHeader("anthropic-version", "2023-06-01")
                .defaultHeader("Content-Type", "application/json")
                .build().post().uri("/v1/messages")
                .bodyValue(body.toString()).retrieve().bodyToMono(String.class).block();
        return extractAnthropicContent(response);
    }

    private String completeAnthropicWithHistory(String system, List<Map<String, String>> history) {
        ObjectNode body = mapper.createObjectNode();
        body.put("model", env.getProperty("llm.model", "claude-sonnet-4-6"));
        body.put("max_tokens", 1024);
        body.put("system", system);
        ArrayNode messages = body.putArray("messages");
        history.forEach(m -> messages.addObject().put("role", m.get("role")).put("content", m.get("content")));
        String response = WebClient.builder().baseUrl("https://api.anthropic.com")
                .defaultHeader("x-api-key", env.getProperty("llm.api-key"))
                .defaultHeader("anthropic-version", "2023-06-01")
                .defaultHeader("Content-Type", "application/json")
                .build().post().uri("/v1/messages")
                .bodyValue(body.toString()).retrieve().bodyToMono(String.class).block();
        return extractAnthropicContent(response);
    }

    // ── Groq fallback embedding (hash-based, not semantic) ───────────────────

    /**
     * Groq and Anthropic don't have embedding APIs.
     * This deterministic hash fallback allows the code to run — but results
     * won't be semantically meaningful. For real RAG, use Azure OpenAI or OpenAI.
     */
    private float[] embedFallback(String text) {
        float[] vec = new float[128];
        for (String word : text.toLowerCase().split("\\s+")) {
            vec[Math.abs(word.hashCode() % 128)] += 1.0f;
        }
        float norm = 0;
        for (float v : vec) norm += v * v;
        norm = (float) Math.sqrt(norm);
        if (norm > 0) for (int i = 0; i < vec.length; i++) vec[i] /= norm;
        return vec;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private String extractContent(String response) {
        try {
            return mapper.readTree(response).path("choices").get(0)
                    .path("message").path("content").asText();
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse LLM response: " + response, e);
        }
    }

    private String extractAnthropicContent(String response) {
        try {
            return mapper.readTree(response).path("content").get(0).path("text").asText();
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse Anthropic response: " + response, e);
        }
    }

    private float[] parseEmbedding(String response) {
        try {
            JsonNode values = mapper.readTree(response).path("data").get(0).path("embedding");
            float[] vec = new float[values.size()];
            for (int i = 0; i < values.size(); i++) vec[i] = (float) values.get(i).asDouble();
            return vec;
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse embedding: " + response, e);
        }
    }
}

File: kapeko-track-a-complete/backend/src/services/llmService.js
// llmService.js — Track A (Node.js)
// Provider-agnostic LLM client
// Same concept as LlmService.java in Track B/C
// Supports: Azure OpenAI | OpenAI | Groq
//
// Switch provider in .env:
//   LLM_PROVIDER=azure | openai | groq

const axios = require('axios');

/**
 * Call the configured LLM and return the text response.
 * @param {string} systemPrompt - The system / role instruction
 * @param {string} userPrompt   - The user message / task
 * @returns {Promise<string>}   - The model's text response
 */
async function complete(systemPrompt, userPrompt) {
  const provider = process.env.LLM_PROVIDER || 'azure';

  switch (provider.toLowerCase()) {
    case 'groq':   return callGroq(systemPrompt, userPrompt);
    case 'openai': return callOpenAI(systemPrompt, userPrompt);
    case 'azure':
    default:       return callAzureOpenAI(systemPrompt, userPrompt);
  }
}

// ── Azure OpenAI ───────────────────────────────────────────────────────────
async function callAzureOpenAI(system, user) {
  return callAzureChat([
    { role: 'system', content: system },
    { role: 'user',   content: user   },
  ]);
}

async function callAzureChat(messages) {
  const endpoint = requireAzureEndpoint();
  const apiKey   = process.env.LLM_API_KEY;
  const isV1Endpoint = endpoint.endsWith('/openai/v1');

  const response = await axios.post(
    `${endpoint}/chat/completions${isV1Endpoint ? '' : '?api-version=2024-02-01'}`,
    {
      ...(isV1Endpoint ? { model: process.env.LLM_MODEL || 'gpt-4o' } : {}),
      messages,
      temperature: 0.7,
    },
    {
      headers: {
        'api-key':      apiKey,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data.choices[0].message.content;
}

function requireAzureEndpoint() {
  const endpoint = (process.env.AZURE_OPENAI_ENDPOINT || '').replace(/\/+$/, '');
  if (!endpoint) {
    throw new Error(
      'AZURE_OPENAI_ENDPOINT is missing. Add it to kapeko-track-a-complete/.env.'
    );
  }
  try {
    new URL(endpoint);
  } catch {
    throw new Error('AZURE_OPENAI_ENDPOINT must be a complete URL starting with https://.');
  }
  return endpoint;
}

// ── OpenAI ─────────────────────────────────────────────────────────────────
async function callOpenAI(system, user) {
  const model  = process.env.LLM_MODEL || 'gpt-4o';
  const apiKey = process.env.LLM_API_KEY;

  const response = await axios.post(
    'https://api.openai.com/v1/chat/completions',
    {
      model,
      messages: [
        { role: 'system', content: system },
        { role: 'user',   content: user   },
      ],
      temperature: 0.7,
    },
    {
      headers: {
        Authorization:  `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data.choices[0].message.content;
}

// ── Groq (free tier — OpenAI-compatible) ───────────────────────────────────
async function callGroq(system, user) {
  const model  = process.env.LLM_MODEL || 'llama-3.1-8b-instant';
  const apiKey = process.env.LLM_API_KEY;

  const response = await axios.post(
    'https://api.groq.com/openai/v1/chat/completions',
    {
      model,
      messages: [
        { role: 'system', content: system },
        { role: 'user',   content: user   },
      ],
      temperature: 0.7,
    },
    {
      headers: {
        Authorization:  `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data.choices[0].message.content;
}

module.exports = { complete };

// ── completeWithHistory() — for Lab 6 Shopping Copilot ───────────────────
async function completeWithHistory(systemPrompt, history) {
  const provider = process.env.LLM_PROVIDER || 'azure';
  const messages = [
    { role: 'system', content: systemPrompt },
    ...history
  ];

  if (provider === 'anthropic') {
    return callAnthropic(systemPrompt, history);
  }

  if (provider !== 'groq' && provider !== 'openai') {
    return callAzureChat(messages);
  }

  const body = {
    model: process.env.LLM_MODEL || (provider === 'groq' ? 'llama-3.1-8b-instant' : 'gpt-4o'),
    temperature: 0.7,
    messages,
  };

  const url = provider === 'groq'
    ? 'https://api.groq.com/openai/v1/chat/completions'
    : 'https://api.openai.com/v1/chat/completions';

  const headers = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${process.env.LLM_API_KEY}` };

  const res = await fetch(url, { method: 'POST', headers, body: JSON.stringify(body) });
  const data = await res.json();
  if (!res.ok || !data.choices) {
    throw new Error(`LLM API error ${res.status}: ${data.error?.message || JSON.stringify(data)}`);
  }
  return data.choices[0].message.content;
}

// ── embed() — for Labs 7 and 8 RAG ───────────────────────────────────────
async function embed(text) {
  const provider = process.env.LLM_PROVIDER || 'azure';

  if (provider === 'groq' || provider === 'anthropic') {
    return embedFallback(text); // No embedding API — hash fallback
  }

  const azureEndpoint = provider === 'azure' ? requireAzureEndpoint() : '';
  const isAzureV1Endpoint = azureEndpoint.endsWith('/openai/v1');
  const url = provider === 'openai'
    ? 'https://api.openai.com/v1/embeddings'
    : `${process.env.AZURE_OPENAI_EMBEDDING_ENDPOINT || azureEndpoint}/embeddings${isAzureV1Endpoint ? '' : '?api-version=2024-02-01'}`;

  const headers = provider === 'azure'
    ? { 'Content-Type': 'application/json', 'api-key': process.env.LLM_API_KEY }
    : { 'Content-Type': 'application/json', 'Authorization': `Bearer ${process.env.LLM_API_KEY}` };

  const body = provider === 'openai'
    ? { model: 'text-embedding-ada-002', input: text }
    : {
        ...(isAzureV1Endpoint
          ? { model: process.env.AZURE_OPENAI_EMBEDDING_DEPLOYMENT || 'text-embedding-3-small' }
          : {}),
        input: text,
      };

  const res = await fetch(url, { method: 'POST', headers, body: JSON.stringify(body) });
  const data = await res.json();
  if (!res.ok || !data.data) {
    throw new Error(`Embedding API error ${res.status}: ${data.error?.message || JSON.stringify(data)}`);
  }
  return data.data[0].embedding;
}

// Hash fallback for Groq (not semantic — allows code to run without errors)
function embedFallback(text) {
  const vec = new Array(128).fill(0);
  const words = text.toLowerCase().split(/\s+/);
  for (const word of words) {
    let hash = 0;
    for (let i = 0; i < word.length; i++) hash = ((hash << 5) - hash) + word.charCodeAt(i);
    vec[Math.abs(hash) % 128] += 1;
  }
  const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0));
  return norm > 0 ? vec.map(v => v / norm) : vec;
}

module.exports = { complete, completeWithHistory, embed };
